backend/src/util/RuntimeManager.ts

```diff
@@ -0,0 +1,189 @@
+  private rcMigratorAvailable = false;
+  private runtimeVersionUnsubscribe?: VoidFn;
+  private pollInterval?: NodeJS.Timeout;
+  private initializationPromise?: Promise<void>;
+
+  private constructor() {}
+
+  public static getInstance(): RuntimeManager {
+    if (!RuntimeManager.instance) {
+      RuntimeManager.instance = new RuntimeManager();
+    }
+    return RuntimeManager.instance;
+  }
+
+  public async initializeRuntimeDetection(): Promise<void> {
+    // Prevent multiple initialization attempts
+    if (this.initializationPromise) {
+      return this.initializationPromise;
+    }
+
+    this.initializationPromise = this.performInitialization();
+    return this.initializationPromise;
+  }
+
+  private async performInitialization(): Promise<void> {
+    const api = await AbstractApi.getInstance().getRelayChainApi();
+    
+    Log.service({
+      service: 'Runtime Manager',
+      action: 'Starting runtime detection initialization'
+    });
+
+    // Initial check
+    if (this.checkPalletAvailability(api)) {
+      this.rcMigratorAvailable = true;
+      this.notifyPalletAvailable();
+
+      return;
+    }
+
+    Log.service({
+      service: 'Runtime Manager',
+      action: 'rcMigrator pallet not available yet, setting up detection'
+    });
+
+    // Subscribe to runtime upgrades
+    try {
+      this.runtimeVersionUnsubscribe = await api.rpc.state.subscribeRuntimeVersion(
+        async (version) => {
+          Log.service({
+            service: 'Runtime Manager',
+            action: 'Runtime version updated',
+            details: { specVersion: version.specVersion.toNumber() }
+          });
+          
+          if (!this.rcMigratorAvailable && this.checkPalletAvailability(api)) {
+            this.rcMigratorAvailable = true;
+            this.notifyPalletAvailable();
+            this.cleanup();
+          }
+        }
+      );
+
+      Log.service({
+        service: 'Runtime Manager',
+        action: 'Successfully subscribed to runtime version updates'
+      });
+    } catch (error) {
+      Log.service({
+        service: 'Runtime Manager',
+        action: 'Failed to subscribe to runtime version, falling back to polling',
+        error: error as Error
+      });
+      
+      // Fallback to polling
+      this.startPolling();
+    }
+  }
+
+  private checkPalletAvailability(api: ApiPromise): boolean {
+    try {
+      return !!(api.query.rcMigrator && api.query.rcMigrator.rcMigrationStage);
+    } catch (error) {
+      // If there's an error accessing the pallet, it's not available
+      return false;
+    }
+  }
+
+  private notifyPalletAvailable(): void {
+    Log.service({
+      service: 'Runtime Manager',
+      action: 'rcMigrator pallet is now available, notifying subscribers'
+    });
+
+    eventService.emit('rcMigratorAvailable', {
+      timestamp: new Date().toISOString()
+    });
+  }
+
+  private async startPolling(): Promise<void> {
+    let attempt = 0;
+    const maxAttempts = 50;
```

Comment:
This is a decent fallback mechanism, but it should not just stop working at 50 attempts. I just picked an arbitrary number, but going to improve on this.